src/hv_aic_events.c

```diff
@@ -0,0 +1,166 @@
+/* SPDX-License-Identifier: MIT */
+
+#include "adt.h"
+#include "hv.h"
+#include "uartproxy.h"
+
+#define AIC_EVENT      0x2004
+#define AIC_EVENT_TYPE GENMASK(31, 16)
+#define AIC_EVENT_NUM  GENMASK(15, 0)
+
+#define AIC_EVENT_TYPE_HW   1
+#define AIC_EVENT_TYPE_IPI  4
+#define AIC_EVENT_IPI_OTHER 1
+#define AIC_EVENT_IPI_SELF  2
+
+#define AIC_MAX_HW_NUM 1024
+
+static u64 aic_base;
+
+static u8 trace_hw_num[AIC_MAX_HW_NUM / 8];
+
+static void set_hw_num(u16 num)
+{
+    u8 *b = &trace_hw_num[num / 8];
+    *b |= 1 << (num & 0x7);
+}
+
+static void clear_hw_num(u16 num)
+{
+    u8 *b = &trace_hw_num[num / 8];
+    *b &= ~(1 << (num & 0x7));
+}
+
+static bool check_hw_num(u16 num)
+{
+    u8 val = trace_hw_num[num / 8];
+    return !!(val & (1 << (num & 0x7)));
+}
+
+static void emit_irqtrace(u16 type, u16 num)
+{
+    struct hv_evt_irqtrace evt = {
+        .flags = 0,
```

Comment:
Let's have a IRQTRACE_IRQ = 1 enum thing here for the flags, marking this as an IRQ event. This is considering we might want to trace mask/unmask operations from the guest in the future. Making this a bitmask keeps it consistent with the configuration function (see below).

Reply:
done